src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -31,83 +33,334 @@ constexpr float DEFAULT_QUEUE_PRIORITY = 1.0f;
+        // shaderUniformBufferArrayDynamicIndexing
+        "arrays of uniform buffers which can be indexed by dynamically uniform integer expressions in shader code",
+        // shaderSampledImageArrayDynamicIndexing
+        "arrays of samplers or sampled images which can be indexed by dynamically uniform integer expressions in "
+        "shader code",
+        // shaderStorageBufferArrayDynamicIndexing
+        "arrays of storage buffers which can be indexed by dynamically uniform integer expressions in shader code",
+        // shaderStorageImageArrayDynamicIndexing
+        "arrays of storage images which can be indexed by dynamically uniform integer expressions in shader code",
+        // shaderClipDistance
+        "clip distances which are supported in shader code",
+        // shaderCullDistance
+        "cull distances in shader code",
+        // shaderFloat64
+        "64-bit floats (doubles) in shader code",
+        // shaderInt64
+        "64-bit integers (signed and unsigned) in shader code",
+        // shaderInt16
+        "16-bit integers (signed and unsigned) in shader code",
+        // shaderResourceResidency
+        "image operations that return resource residency information in shader code",
+        // shaderResourceMinLod
+        "image operations specifying the minimum resource LOD in shader code",
+        // sparseBinding
+        "resource memory which can be managed at opaque sparse block level instead of at the object level",
+        // sparseResidencyBuffer
+        "access to partially resident buffers",
+        // sparseResidencyImage2D
+        "access to partially resident 2D images with 1 sample per pixel",
+        // sparseResidencyImage3D
+        "access to partially resident 3D images",
+        // sparseResidency2Samples
+        "access to partially resident 2D images with 2 samples per pixel",
+        // sparseResidency4Samples
+        "access to partially resident 2D images with 4 samples per pixel",
+        // sparseResidency8Samples
+        "access to partially resident 2D images with 8 samples per pixel",
+        // sparseResidency16Samples
+        "access to partially resident 2D images with 16 samples per pixel",
+        // sparseResidencyAliased
+        "correct access to data aliased into multiple locations",
+        // variableMultisampleRate
+        "variable multisample rate",
+        // inheritedQueries
+        "execution of secondary command buffers while a query is active"};
+
+    if (index > feature_descriptions.size()) {
+        return "";
+    }
+    return feature_descriptions[index];
+}
 
-    std::vector<VkExtensionProperties> device_extensions(device_extension_count);
+/// Get the name of a physical device
+/// @param physical_device The physical device
+/// @return The name of the physical device
+std::string get_physical_device_name(const VkPhysicalDevice physical_device) {
```

Comment:
Not used in device constructor.

Reply:
It is used! See line 357:

```cpp
for (std::size_t i = 0; i < FEATURE_COUNT; i++) {
    if (comparable_required_features[i] == VK_TRUE) {
        features_to_enable[i] = VK_TRUE;
    }
    if (comparable_optional_features[i] == VK_TRUE && comparable_available_features[i] == VK_TRUE) {
        features_to_enable[i] = VK_TRUE;
    }
    if (comparable_optional_features[i] == VK_TRUE && comparable_available_features[i] == VK_FALSE) {
        spdlog::warn("The physical device {} does not support {}!", get_physical_device_name(physical_device),
                     vk_tools::get_device_feature_description(i));
    }
}
```

I would like to keep that function in `device.hpp`

Reply:
Thats line 472?
Iam speaking of line 358-361

Reply:
I think line numbers are outdated by now, since I added so many changes?

Reply:
I am using it everywhere now.